File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head is None :
            return None
        
        nodes=[]
        cur=head
        while cur:
            nodes.append(cur)
            cur=cur.next
        length=len(nodes)
        if length==n:
            return head.next
        nodes[length-n-1].next=nodes[length-n].next if (length -n) <length else None
        return head
        
        

File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode s = head  , f = head;
        while(f!=null && f.next!=null){
            s= s.next;
            f=f.next.next;
        }

        ListNode curr = s.next;
        s.next=null;
        ListNode prev = null;
        while(curr!=null){
            ListNode temp  = curr.next;
            curr.next=prev;
            prev=curr;
            curr=temp;

        }
        ListNode first = head;
        ListNode second = prev;
        while (second != null) {
            ListNode tmp1 = first.next;
            ListNode tmp2 = second.next;
            first.next = second;
            second.next = tmp1;
            first = tmp1;
            second = tmp2;
        }
    }
}
